File: android/src/main/java/it/hubagency/spatialscan/FloorPlaneAnchor.kt
package it.hubagency.spatialscan

import com.google.ar.core.Plane
import com.google.ar.core.TrackingState

/**
 * Stabilizza la stima di floorY usando mediana mobile su campioni stabili.
 *
 * Logica:
 *  1. Raccoglie campioni di floorY da floor plane ARCore (o fallback camera)
 *  2. Lock quando la mediana degli ultimi 30 campioni ha varianza < 0.5cm²
 *     E il tracking è stato TRACKING continuo
 *  3. Una volta bloccato, aggiorna la mediana con peso basso (aggiornamento lento)
 *
 * Fallback: cameraPoseY - 1.2f usato SOLO come filtro altezza point cloud,
 * non per coordinate definitive dei muri.
 */
class FloorPlaneAnchor {

    private val samples      = ArrayDeque<Float>(60)
    private var lockedFloorY : Float? = null
    private var trackingStreak = 0    // frame consecutivi con TRACKING

    val isLocked: Boolean get() = lockedFloorY != null

    /**
     * @param planes        tutti i planes del frame (per trovare il floor)
     * @param cameraY       Y world della camera (fallback)
     * @param isTracking    se il tracking è attivo in questo frame
     * @return              stima stabile di floorY
     */
    fun update(planes: Collection<Plane>, cameraY: Float, isTracking: Boolean): Float {
        if (!isTracking) {
            trackingStreak = 0
            return lockedFloorY ?: (cameraY - 1.2f)
        }
        trackingStreak++

        // Trova il floor plane (il piano orizzontale più grande)
        val floor = planes
            .filter { it.trackingState == TrackingState.TRACKING && it.subsumedBy == null }
            .filter { it.type == Plane.Type.HORIZONTAL_UPWARD_FACING }
            .maxByOrNull { it.extentX * it.extentZ }

        val raw = floor?.centerPose?.ty() ?: (cameraY - 1.2f)

        // Accumula campioni solo se tracking stabile (evita outlier da pause brevi)
        if (trackingStreak > 5) {
            if (samples.size >= 60) samples.removeFirst()
            samples.addLast(raw)
        }

        // Prova a bloccare: serve varianza bassa su 30+ campioni con tracking continuo
        if (lockedFloorY == null && samples.size >= 30 && trackingStreak >= 30) {
            val med = median(samples)
            val variance = samples.map { (it - med) * (it - med) }.average().toFloat()
            if (variance < 0.0005f) {   // std < ~2.2cm
                lockedFloorY = med
            }
        }

        // Aggiornamento molto lento del lock — solo se la mediana diverge > 5cm
        // Evita drift da jitter camera. EMA 0.002 = ~500 frame per adattarsi a 1cm.
        lockedFloorY?.let { locked ->
            val med = median(samples)
            if (kotlin.math.abs(med - locked) > 0.05f) {
                lockedFloorY = locked * 0.998f + med * 0.002f
            }
        }

        return lockedFloorY ?: raw
    }

    fun reset() {
        samples.clear()
        lockedFloorY = null
        trackingStreak = 0
    }

    private fun median(list: Collection<Float>): Float {
        val sorted = list.sorted()
        val n = sorted.size
        return if (n % 2 == 0) (sorted[n / 2 - 1] + sorted[n / 2]) / 2f else sorted[n / 2]
    }
}

File: ios/Plugin/FloorPlaneAnchor.swift
import ARKit

class FloorPlaneAnchor {

    private var samples       = [Float]()       // buffer max 60 campioni
    private var lockedFloorY: Float?
    private var trackingStreak = 0              // frame consecutivi TRACKING

    var isLocked: Bool { lockedFloorY != nil }

    func update(frame: ARFrame, isTracking: Bool) -> Float {
        let cameraY = frame.camera.transform.columns.3.y

        guard isTracking else {
            trackingStreak = 0
            return lockedFloorY ?? (cameraY - 1.2)
        }
        trackingStreak += 1

        let planes = frame.anchors.compactMap { $0 as? ARPlaneAnchor }
        let horizontalPlanes = planes.filter { $0.alignment == .horizontal }
        let floor: ARPlaneAnchor? = horizontalPlanes.max(by: {
            ($0.extent.x * $0.extent.z) < ($1.extent.x * $1.extent.z)
        })
        let raw: Float = floor.map { Float($0.transform.columns.3.y) } ?? (cameraY - 1.2)

        if trackingStreak > 5 {
            if samples.count >= 60 { samples.removeFirst() }
            samples.append(raw)
        }

        if lockedFloorY == nil && samples.count >= 30 && trackingStreak >= 30 {
            let med      = median(samples)
            let variance = samples.map { ($0 - med) * ($0 - med) }.reduce(0, +) / Float(samples.count)
            if variance < 0.0005 {
                lockedFloorY = med
            }
        }

        if let locked = lockedFloorY {
            let med = median(samples)
            if abs(med - locked) > 0.05 {
                lockedFloorY = locked * 0.998 + med * 0.002
            }
        }

        return lockedFloorY ?? raw
    }

    func reset() {
        samples.removeAll()
        lockedFloorY = nil
        trackingStreak = 0
    }

    private func median(_ list: [Float]) -> Float {
        let sorted = list.sorted()
        let n = sorted.count
        return n % 2 == 0 ? (sorted[n / 2 - 1] + sorted[n / 2]) / 2 : sorted[n / 2]
    }
}
